Rewrite of non-commutative multiplication matching

anpandey opened the pull request:
#### References to other Issues or PRs

Fixes #17172

#### Brief description of what is fixed or changed

Rewrites the matching code for non-commutative multiplication expressions that:

* Respects expression structure (`A*w`matches`A*B` but not`C*B`)
* Accounts for non-commutativity (`A*B*w` matches `A*B*C` but not `B*A*C`)
* Enables non-commutative wildcards to match any subsequence (`A*w*D` matches `A*B*C*D`)

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* core
  * improvements to non-commutative matching
<!-- END RELEASE NOTES -->

anpandey committed:
Rewrite of non-commutative matching

Rewrite of matching of non-commutative expressions that:
- Respects expression structure (A*w matches A*B but not C*B)
- Accounts for non-commutativity (A*B*w matches A*B*C but not B*A*C)
- Enables non-commutative wildcards to match any subsequence (A*w*D
  matches A*B*C*D)

anpandey committed:
Add additional non-commutative tests

certik commented:
Does this have performance implications? I don't know if this method is used in the core or not.

anpandey commented:
I don't think it should affect the performance of anything else in the core. It seems like `match` is only every used for commutative arguments, in which case the performance should be identical.

asmeurer commented:
Great. How hard is it to now build matrix expression matching on top of this?

anpandey commented:
@asmeurer It should be enough to define the matrix wildcard class and its `match` method. The matching code in `Mul` itself should take care of everything else.

anpandey committed:
Two more assertions

anpandey committed:
Use indices in wildcard dictionary

anpandey committed:
Check if matches contradict eachother

If a wildcard is more than once in a pattern, make sure that it matches
the same subexpression every time

anpandey committed:
Add mismatch and positive exponent tests

anpandey committed:
Some docstring and comment updates

anpandey committed:
Rename `matpows` to just `pows`
